File: public/tools/elements/app/GtTitleBar.ts
import {PolymerElement} from "../../polymer/PolymerElement";
import {Element, Property, Watch, Listener, Inject, On} from "../../polymer/Annotations";
import {GtButton} from "../widgets/GtButton";
import {ViewMetadata, Tab} from "../../client/router/View";
import {Router} from "../../client/router/Router";
import {Server} from "../../client/server/Server";
import {ChatService} from "../../services/chat/ChatService";

@Element({
	selector: "gt-title-bar",
	template: "/assets/imports/app.html",
	dependencies: [GtButton]
})
export class GtTitleBar extends PolymerElement {
	// Remove the title bar if not launched as an app
	private ready() {
		if (!this.app.standalone)
			this.$["window-controls"].remove();
	}

	// ========================================================================
	// Loading

	@Inject
	@Watch({ loading: "UpdateLoading" })
	private server: Server;

	@Property
	private loading: boolean;

	private UpdateLoading() {
		this.loading = this.server.loading;
	}

	// ========================================================================
	// Chat

	@Inject
	@On({
		"connected": "UpdateOnlineCount",
		"disconnected": "UpdateOnlineCount"
	})
	private chat: ChatService;

	@Property
	public online_users: number = 0;

	private UpdateOnlineCount() {
		this.debounce("update-online", () => this.online_users = this.chat.onlinesUsers.length);
	}

	// ========================================================================
	// Tabs

	@Inject
	@On({ "route-updated": "UpdateTabs" })
	private router: Router;

	private tabs: Tab[];

	private async UpdateTabs() {
		// Get tabs generator from view metadata
		let meta = Reflect.getMetadata<ViewMetadata>("view:meta", this.router.activeView);
		if (!meta) return;

		this.tabs = meta.tabs(this.router.activeView, this.router.activePath, this.app.user).filter(t => !t.hidden);
	}

	// ========================================================================
	// Panel

	@Property({ reflect: true })
	public panel: boolean = false;

	@Listener("logo.click")
	private OpenPanel() {
		if (!this.panel) {
			this.panel = true;
			(<any>this).style.zIndex = 20;
		}
	}

	@Listener("panel.click")
	private PanelClicked(ev: MouseEvent) {
		if (this.panel) {
			ev.stopImmediatePropagation();
			ev.preventDefault();
			this.ClosePanel();
		}
	}

	@Listener("panel.mouseleave")
	private ClosePanel() {
		if (this.panel) {
			this.panel = false;
			this.debounce("z-index-downgrade", () => { if (!this.panel) (<any>this).style.zIndex = 10; }, 300);
		}
	}

	private Reload() {
		document.location.reload();
	}

	private DownloadClient() { }

	private DevTools() { }

	private Logout() { }

	private Quit() {
		window.close();
	}
}

File: client/src/gt/component/app/GtTitleBar.scala
package gt.component.app

import gt.component.widget.form.GtButton
import gt.component.{GtHandler, Tab}
import gt.{Router, Server}
import org.scalajs.dom.{MouseEvent, document, window}
import rx.Var
import scala.scalajs.js.timers.setTimeout
import util.jsannotation.js
import xuen.Component

object GtTitleBar extends Component[GtTitleBar](
	selector = "gt-title-bar",
	templateUrl = "/assets/imports/app.html",
	dependencies = Seq(GtButton)
)

@js class GtTitleBar extends GtHandler {
	override def ready(): Unit = {
		// Remove the title bar if not launched as an app
		if (!app.standalone) {
			val ctrls = child.`window-controls`
			ctrls.parentNode.removeChild(ctrls)
		}
	}

	// ==========================
	// Loader

	// Track loading status
	val loading = Server.loading

	// ==========================
	// Tabs

	// Current visible tabs
	val tabs = Var(Nil: Seq[Tab])
	Router.onUpdate ~> { case (path, _, view) =>
		tabs := view.tabs(view.selector, path, app.user).filter(!_.hidden)
	}

	// Number of online players
	var onlines = Var(0)

	// ==========================
	// Panel

	// Panel status
	val panel = attribute[Boolean] := false
	var panelDebounce = false

	/** Opens the panel */
	def openPanel(): Unit = if (!panel) {
		panel := true
		style.zIndex = "20"
	}

	/** Absorbs clicks on the panel */
	def panelClicked(ev: MouseEvent): Unit = if (panel) {
		ev.stopImmediatePropagation()
		ev.preventDefault()
		closePanel()
	}

	/** Closes panel on mouseleave */
	def closePanel(): Unit = if (panel) {
		panel := false
		if (panelDebounce) {
			panelDebounce = true
			setTimeout(300) {
				if (!panel) style.zIndex = "10"
				panelDebounce = false
			}
		}
	}

	def reload(): Unit = document.location.reload()
	def downloadClient(): Unit = {}
	def devTools(): Unit = {}
	def logout(): Unit = {}
	def quit(): Unit = window.close()
}
